x/slashing/keeper/infractions.go

```diff
@@ -36,30 +36,105 @@ func (k Keeper) HandleValidatorSignatureConcurrent(ctx sdk.Context, addr cryptot
 
 	window := k.SignedBlocksWindow(ctx)
 
+	index := signInfo.IndexOffset % k.SignedBlocksWindow(ctx)
+
 	missedInfo, found = k.GetValidatorMissedBlocks(ctx, consAddr)
+
 	if !found {
+		arrLen := window / 64
+		if window%64 > 0 {
+			arrLen += 1
+		}
 		missedInfo = types.ValidatorMissedBlockArray{
-			Address:       consAddr.String(),
-			MissedHeights: make([]int64, 0),
+			Address:      consAddr.String(),
+			WindowSize:   window,
+			MissedBlocks: make([]uint64, arrLen),
 		}
 	}
-	cutoff := height - window
-	// go through and increment number evicted as necessary
-	numberEvicted := 0
-	for _, missedHeight := range missedInfo.MissedHeights {
-		if missedHeight <= cutoff {
-			numberEvicted += 1
-		} else {
-			break
+	// TODO:  resize if necessary?
+	if found && missedInfo.WindowSize != window {
+		// we need to resize the missed block array AND update the signing info accordingly
+		switch {
+		case missedInfo.WindowSize < window:
+			// missed block array too short, lets expand it
+			boolArray := k.ParseBitGroupsToBoolArray(missedInfo.MissedBlocks, missedInfo.WindowSize)
+			oldIndex := signInfo.IndexOffset % missedInfo.WindowSize
+			newArray := make([]bool, window)
+			copy(newArray[0:oldIndex+1], boolArray[0:oldIndex+1])
+			if oldIndex+1 < missedInfo.WindowSize {
+				copy(newArray[oldIndex+1+(window-missedInfo.WindowSize):], boolArray[oldIndex+1:])
+			}
+			// rotate array from oldIndex to newIndex
+			shiftDiff := ((index - oldIndex) + window) % window
+			finalBoolArr := make([]bool, window)
+			copy(finalBoolArr, newArray[window-shiftDiff:])
+			copy(finalBoolArr[shiftDiff:], newArray[:window-shiftDiff])
+			missedInfo.MissedBlocks = k.ParseBoolArrayToBitGroups(finalBoolArr)
+			missedInfo.WindowSize = window
+			// TODO: set missed blocks later
+			// k.SetValidatorMissedBlocks(ctx, writeInfo.ConsAddr, missedInfo)
+		case missedInfo.WindowSize > window:
+			// missed block array too long, we need to trim
+			// we need to keep the last N blocks prior to the validator index offset (wrapping around backwards if necessary)
```

Comment:
IMO Shrinking window size is a risky operation that might cause lot of validators to be jailed at the same time when window size change if here we decided to keep the history.

My recommendation (not sure if make sense or not) is to keep things simple, let's just drop all the history when the window size is shrinked, recreate a new empty array list to start with. This window size change should be pretty rare to happen and it's actually safer to if we missed to punish validators than we suddenly started to jail a bunch of validators.

Reply:
Yeah this is reasonable, although ideally vals shouldn't be jailed because we update missed counters to exclude missed blocks from outside the window